### PR TITLE
[export] Allow modules to be created in the forward

### PR DESCRIPTION
**Pull request by angelayi**
Fixes the error in non-strict export when we're tracing a module that initializes another module in its forward function. This appears in [many huggingface models](https://github.com/search?q=repo%3Ahuggingface%2Ftransformers+CrossEntropyLoss%28%29&type=code&fbclid=IwAR285uKvSevJM6SDbXmb4-monj4iH7wf8opkvnec-li7sKpn4lUMjIvbGKc). It's probably not good practice to do this, but since it appears in so many places, and strict-export supports this, we will also support this. 

The approach we'll take for these cases is that we will inline the call to the module. Parameters and buffers initialized as constants (with `torch.tensor`) will be represented as constant tensors, and those initialized with tensor factory functions (`torch.ones`) will show up as an operator in the graph. The module stack for the ops in the inlined module will reflect the toplevel's module stack.

One issue is that strict-export seems to segfault when there is an `nn.Parameter` call in the constructor (https://github.com/pytorch/pytorch/issues/126109). Non-strict export will succeed.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

### CODE REVIEW

File: test/export/test_export.py [hunk 1]
```diff
@@ -3249,6 +3249,82 @@ def dynamify_inp(x):
...
+            def forward(self, x):
+                return x + x
+
+        class M2(torch.nn.Module):
+            def forward(self, x):
+                m = M1()
+                return m(x) * x
+
+        inps = (torch.randn(3, 3),)
+        ep = export(M2(), inps)
```
Comment (ydwu4) [inline test/export/test_export.py:3263]: How about we also assertExpectedInline of the graph? It's ok if strict-mode and non-strict mode has different graph, we can branch on it so that we can clearly see the difference. It's not immediate to see what the graphs look like and what's the difference between strict and nonstrict mode.

File: test/export/test_export.py [hunk 2]
```diff
@@ -3249,6 +3249,82 @@ def dynamify_inp(x):
...
+
+        unflattened = unflatten(ep)
+        self.assertTrue(torch.allclose(unflattened(*inps), M2()(*inps)))
+
+    @testing.expectedFailureRetraceability  # Retracing tensor constants results in buffers
+    def test_nested_module_with_constant_buffer(self):
+        class M1(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.b = torch.tensor(5)
```
Comment (ydwu4) [inline test/export/test_export.py:3308]: Can we also support parameters? If not, adding an expected failure test will also be good.

File: torch/fx/experimental/proxy_tensor.py [hunk 3]
```diff
@@ -1082,7 +1083,17 @@ def call_module(self, m, forward, args, kwargs):
         # use cases don't need to work with HOO.
         if isinstance(m, (OptimizedModule, GraphModule)):
             return forward(*args, **kwargs)
-        return Tracer.call_module(self, m, forward, args, kwargs)
+
+        try:
```
Comment (ydwu4) [inline torch/fx/experimental/proxy_tensor.py:1087]: If it's possible, we should restrict the try except block to the minimal. NameError has a larger blast radius. We could wrap the path_of_module call by overriding the path_of_module and raise a new exception to catch here:
```
def path_of_module(self...):
  try:
    super().path_of_module(...)
  except NameError as e:
    raise InitializingModInForwardError(...)
```

File: test/export/test_export.py (hunk 2 above)
Comment (angelayi) [reply test/export/test_export.py:3329]: Parameters in strict-export cause a segfault, so I added a test for only using nonstrict-export. I'm not sure how to expect failure the strict-export case w/o expect failure-ing the generated nonstrict test.

File: test/export/test_export.py [hunk 4]
```diff
@@ -3249,6 +3249,158 @@ def dynamify_inp(x):
...
+        self.assertEqual(len(ep.state_dict), 0)
+        self.assertEqual(len(ep.constants), 1)
+
+        self.assertExpectedInline(
+            str(ep.graph).strip(),
+            """\
+graph():
+    %c_lifted_tensor_0 : [num_users=1] = placeholder[target=c_lifted_tensor_0]
+    %x : [num_users=2] = placeholder[target=x]
+    %lift_fresh_copy : [num_users=1] = call_function[target=torch.ops.aten.lift_fresh_copy.default](args = (%c_lifted_tensor_0,), kwargs = {})
```
Comment (ydwu4) [inline test/export/test_export.py:3352]: Do we know where this copy-then-detach pattern is added?

**Review by ydwu4 — APPROVED** [review]

File: test/export/test_export.py (hunk 4 above)
Comment (angelayi) [reply test/export/test_export.py:3352]: I think it's from the `torch.tensor(5)` creation.

**Comment by angelayi** [comment]
CI fix here: https://github.com/pytorch/pytorch/pull/126715
